File: patterns/function-calling/openai-like/python/basic_tools.py
import os
import json
from openai import OpenAI
from dotenv import load_dotenv

load_dotenv()

def get_current_weather(location, unit="celsius"):
    """Get the current weather in a given location"""
    if "tokyo" in location.lower():
        return json.dumps({"location": "Tokyo", "temperature": "10", "unit": unit})
    elif "san francisco" in location.lower():
        return json.dumps({"location": "San Francisco", "temperature": "72", "unit": "fahrenheit"})
    elif "paris" in location.lower():
        return json.dumps({"location": "Paris", "temperature": "22", "unit": unit})
    else:
        return json.dumps({"location": location, "temperature": "unknown"})

def main():
    client = OpenAI(api_key=os.environ.get("OPENAI_API_KEY"))

    # Tool definition
    tools = [
        {
            "type": "function",
            "function": {
                "name": "get_current_weather",
                "description": "Get the current weather in a given location",
                "parameters": {
                    "type": "object",
                    "properties": {
                        "location": {
                            "type": "string",
                            "description": "The city and state, e.g. San Francisco, CA",
                        },
                        "unit": {"type": "string", "enum": ["celsius", "fahrenheit"]},
                    },
                    "required": ["location"],
                },
            },
        }
    ]

    messages = [{"role": "user", "content": "What's the weather like in San Francisco, Tokyo, and Paris?"}]

    print("Sending request with tools...")
    
    response = client.chat.completions.create(
        model="gpt-3.5-turbo",
        messages=messages,
        tools=tools,
        tool_choice="auto", 
    )

    response_message = response.choices[0].message
    tool_calls = response_message.tool_calls
    
    if tool_calls:
        print(f"Model decided to call {len(tool_calls)} tool(s).")
        
        # Extend conversation with assistant's reply
        messages.append(response_message)
        
        # Execute tools and append their outputs
        for tool_call in tool_calls:
            function_name = tool_call.function.name
            function_args = json.loads(tool_call.function.arguments)
            
            if function_name == "get_current_weather":
                print(f"Calling tool: {function_name} with args: {function_args}")
                function_response = get_current_weather(
                    location=function_args.get("location"),
                    unit=function_args.get("unit"),
                )
                
                messages.append(
                    {
                        "tool_call_id": tool_call.id,
                        "role": "tool",
                        "name": function_name,
                        "content": function_response,
                    }
                )
        
        print("Sending follow-up request to get final answer...")
        second_response = client.chat.completions.create(
            model="gpt-3.5-turbo",
            messages=messages,
        )
        print("\nFinal Answer:")
        print(second_response.choices[0].message.content)

if __name__ == "__main__":
    main()

File: patterns/function-calling/openai-like/javascript/basic-tools.ts
import 'dotenv/config';
import OpenAI from 'openai';

// Mock function
function getCurrentWeather(location: string, unit: string = 'celsius') {
    if (location.toLowerCase().includes('tokyo')) {
        return JSON.stringify({ location: 'Tokyo', temperature: '10', unit: unit });
    } else if (location.toLowerCase().includes('san francisco')) {
        return JSON.stringify({ location: 'San Francisco', temperature: '72', unit: 'fahrenheit' });
    } else if (location.toLowerCase().includes('paris')) {
        return JSON.stringify({ location: 'Paris', temperature: '22', unit: unit });
    } else {
        return JSON.stringify({ location: location, temperature: 'unknown' });
    }
}

async function main() {
    const client = new OpenAI({
        apiKey: process.env.OPENAI_API_KEY,
    });

    const tools: OpenAI.Chat.Completions.ChatCompletionTool[] = [
        {
            type: 'function',
            function: {
                name: 'get_current_weather',
                description: 'Get the current weather in a given location',
                parameters: {
                    type: 'object',
                    properties: {
                        location: {
                            type: 'string',
                            description: 'The city and state, e.g. San Francisco, CA',
                        },
                        unit: { type: 'string', enum: ['celsius', 'fahrenheit'] },
                    },
                    required: ['location'],
                },
            },
        },
    ];

    const messages: OpenAI.Chat.Completions.ChatCompletionMessageParam[] = [
        { role: 'user', content: "What's the weather like in San Francisco, Tokyo, and Paris?" },
    ];

    console.log('Sending request with tools...');

    const response = await client.chat.completions.create({
        model: 'gpt-3.5-turbo',
        messages: messages,
        tools: tools,
        tool_choice: 'auto',
    });

    const responseMessage = response.choices[0].message;
    const toolCalls = responseMessage.tool_calls;

    if (toolCalls) {
        console.log(`Model decided to call ${toolCalls.length} tool(s).`);

        messages.push(responseMessage);

        for (const toolCall of toolCalls) {
            const functionName = toolCall.function.name;
            const functionArgs = JSON.parse(toolCall.function.arguments);

            if (functionName === 'get_current_weather') {
                console.log(`Calling tool: ${functionName} with args:`, functionArgs);

                const functionResponse = getCurrentWeather(
                    functionArgs.location,
                    functionArgs.unit
                );

                messages.push({
                    tool_call_id: toolCall.id,
                    role: 'tool',
                    content: functionResponse,
                });
            }
        }

        console.log('Sending follow-up request to get final answer...');
        const secondResponse = await client.chat.completions.create({
            model: 'gpt-3.5-turbo',
            messages: messages,
        });

        console.log('\nFinal Answer:');
        console.log(secondResponse.choices[0].message.content);
    }
}

main();
